search-service/src/main/kotlin/com/egm/stellio/search/web/TemporalApiResponseBuilder.kt

```diff
@@ -0,0 +1,241 @@
+        )
+        val temporalQuery = query.temporalQuery
+        if (temporalQuery.isLastNInsideLimit()) {
+            return successResponse
+        }
+
+        val attributesWhoReachedLimit = getAttributesWhoReachedLimit(entities, query)
+
+        if (attributesWhoReachedLimit.isEmpty()) {
+            return successResponse
+        }
+
+        val range = getTemporalPaginationRange(attributesWhoReachedLimit, query)
+
+        val filteredEntities = entities.map { filterEntityInRange(it, range, query) }
+        return ResponseEntity.status(HttpStatus.PARTIAL_CONTENT).apply {
+            this.headers(
+                successResponse.headers
+            )
+            this.header(
+                HttpHeaders.CONTENT_RANGE,
+                getHeaderRange(range, query.temporalQuery)
+            )
+        }.body(serializeObject(filteredEntities.toFinalRepresentation(ngsiLdDataRepresentation)))
+    }
+
+    fun buildEntityTemporalResponse(
+        entity: CompactedEntity,
+        query: TemporalEntitiesQuery,
+        mediaType: MediaType,
+        requestParams: MultiValueMap<String, String>,
+        contexts: List<String>,
+    ): ResponseEntity<String> {
+        val ngsiLdDataRepresentation = parseRepresentations(requestParams, mediaType)
+        val successResponseHeader = prepareGetSuccessResponseHeaders(mediaType, contexts)
+
+        if (query.temporalQuery.isLastNInsideLimit()) {
+            return successResponseHeader.body(
+                serializeObject(
+                    entity.toFinalRepresentation(ngsiLdDataRepresentation)
+                )
+            )
+        }
+
+        val attributesWhoReachedLimit = getAttributesWhoReachedLimit(listOf(entity), query)
+
+        if (attributesWhoReachedLimit.isEmpty()) {
+            return successResponseHeader.body(
+                serializeObject(
+                    entity.toFinalRepresentation(ngsiLdDataRepresentation)
+                )
+            )
+        }
+
+        val range = getTemporalPaginationRange(attributesWhoReachedLimit, query)
+        val filteredEntity = filterEntityInRange(entity, range, query)
+
+        return ResponseEntity.status(HttpStatus.PARTIAL_CONTENT)
+            .apply {
+                this.header(
+                    HttpHeaders.CONTENT_RANGE,
+                    getHeaderRange(range, query.temporalQuery)
+                )
+                this.headers(
+                    successResponseHeader.body("").headers
+                )
+            }.body(
+                serializeObject(
+                    filteredEntity.toFinalRepresentation(ngsiLdDataRepresentation)
+                )
+            )
+    }
+
+    private fun getAttributesWhoReachedLimit(entities: List<CompactedEntity>, query: TemporalEntitiesQuery): List<Any> {
+        val temporalQuery = query.temporalQuery
+
+        return entities.flatMap { entity ->
+            entity.values.mapNotNull { attr ->
+
+                when (attr) {
```

Comment:
instead of doing the `when` around the type of the value of the attribute, more robust to do it on the kind of temporal query (default, temporal or aggregated).

in all cases, take care of the datasetId (in default representation, the count must be grouped by datasetId, in other representations, the count must be done for each datasetId)

Reply:
I develop this condition with in mind that we could have temporal attributes mix with non temporal attributes. (Is it the case?)
If this is not the case it may be more logic to do it using the value of attribute.
Else it may make the code even more complicated since i must do this same "when" inside of the new condition.